### ISSUE TITLE
Automatic linkage based on members without having to specify relationship when building document

### ISSUE THREAD

**Issue by cmeeren** [scott-mcdonald/JsonApiFramework#47]
(Forgive any syntax errors, I'm using F# and my C# skills are a bit rusty at the moment.)

Great library! Just got started with it. I have a possible feature request. Consider the following two resources:

```c#
class User : IResource
{
  public string Username { get; set }
  public string FullName { get; set }
}

class Order : IResource
{
  public string OrderId { get; set }
  public User SalesPerson { get; set }
  public string SomeOrderProp { get; set }
}
```

Note that the `Order` class directly references a `User` through one of its properties instead of using a foreign key. I'd like JsonApiFramework to pick up on this and automatically generate relationships without me having to explicitly specify the relationship when building the document in the controller. (Ideally also as a convention without me having to define the relation in a `ResourceTypeBuilder`, but I can live with that if necessary - mostly I want to avoid the linkage boilerplate when building documents.)

For example, given the following instances:

```c#
var order = new Order
{
  OrderNumber = "ABC123",
  SomeOrderProp = "foobar",
  SalesPerson = new Person {Username = "someuser", FullName = "Some User"}
}
```

and the following document creation code:

```c#
documentContext
  .NewDocument(currentRequestUrl)
    .Resource(order)
    .ResourceEnd()
  .WriteDocument()
```

I want the following JSON:

```json
{
    "data": {
        "type": "orders",
        "id": "ABC123",
        "attributes": {
            "some-order-prop": "foobar"
        }
        "relationships": {
            "sales-person": {
                "links": {
                    "self": "...",
                    "related": "..."
                },
                "data": {
                    "type": "users",
                    "id": "someuser"
                }
            }
        }
    }
}
```

I realize of course that there are some decisions here that need to be taken automatically, such as which links to include in the relationship. Unsure if that's a good idea. But still, I figured I could at least make a suggestion. What do you think?

**Comment by scott-mcdonald** [comment]
Christer,

A couple of comments.

First for reference purposes the json:api specification does not allow resources to be composed of other resources, click [here](http://discuss.jsonapi.org/t/clarification-on-resources-composed-of-other-resources/252) to read a whole discussion on the json:api forum if interested. Instead a resource will have relationships to other related resources using "resource linkage".

Based on that constraint by the json:api specification, Order will not automatically contain a related sales person User upon serialization or deserialization even if you modeled it that way in your service model like you did in your example. But the document context does implement the IDocumentReader abstraction where client code can get the "included" related sales person User.

So currently with the JsonApiFramework you would have to do something like the following:

````cs

public class User : IResource
{
    public string UserId { get; set; }
    public string UserName { get; set; }
    public string FullName { get; set; }
}

public class Order : IResource
{
    public string OrderId { get; set; }
    public User SalesPerson { get; set; }
}

public class OrderConfigurationWithConventions : ResourceTypeBuilder<Order>
{
    public OrderConfigurationWithConventions()
    {
        // Attributes to Ignore
        this.Attribute(x => x.SalesPerson).Ignore();
        
        // Relationships
        this.ToOneRelationship<User>("sales-person");
    }
}

````

So on the server-side you would build this document like the following:

````cs

GET http://example.com/orders/1234

public class OrdersController : ApiController
{
    [Route("orders/{id}")]
    public async Task<IHttpActionResult> GetAsync(string id)
    {
        Contract.Requires(String.IsNullOrWhitespace(id) == false);

        // Get order which contains the related sales person
        var order = await GetOrder(id);
        var salesPerson = order.SalesPerson;

        // Build and return JSON API document ////////////////////////////// 
        var currentRequestUrl = HttpContext.Current.Request.Url;
        using (var documentContext = new OrdersDocumentContext())
        {
            // Build new document.
            var document = documentContext
                .NewDocument(currentRequestUrl)

                    // Resource document
                    .Resource(order)
                        // order relationships
                        .Relationships()
                            // order -> sales-person
                            .AddRelationship("sales-person", new [] { "related" })
                        .RelationshipsEnd()

                    .ResourceEnd()

                    .Included()
                        .Include(ToOneIncludedResource.Create(order, "sales-person", salesPerson))
                        .IncludeEnd()
                    .IncludedEnd()

                .WriteDocument();
    
            // Return 200 OK
            // Note: WebApi JsonMediaTypeFormatter serializes the JSON API document into JSON. 
            return this.Ok(document);
        }
    }
}

````

And on the client-side you would read the document like the following:

````cs

public class OrderReader
{
    public void ReadJsonApiDocument(string json)
    {
        // Parse and read JSON API document. 
        var document = JsonObject.Parse<Document>(json);
        using (var documentContext = new OrdersDocumentContext(document))
        {
            // Order
            var order = documentContext.GetResource<Order>();
            var orderRelationships = documentContext.GetResourceRelationships(order);

            // Related Sales Person
            var salesPersonRelationship = articleRelationships ["sales-person"];

            Assume(salesPersonRelationship.IsResourceLinkageNullOrEmpty() == false);
            var salesPerson = documentContext.GetRelatedResource<User>(salesPersonRelationship);
        }
    }
}

````

So with that said as a background, yes I understand you would like to avoid the explicit 

````cs

.Included()
    .Include(ToOneIncludedResource.Create(order, "sales-person", salesPerson))
    .IncludeEnd()
.IncludedEnd()

````

from above and have JsonApiFramework infer based on resource composition but because resource composition is not allowed per the json:api specification I do not plan to put that in anytime soon. That is the bad news...

The good news this will be very possible with ApiFramework and that is what I am focusing on currently. I am running out of time at the moment but will comment about ApiFramework when I can and how this type of scenario is basically in the wheel house of ApiFramework...

**Comment by cmeeren** [comment]
Thank you for the response.

For clarification:

> but because resource composition is not allowed per the json:api specification I do not plan to put that in anytime soon

Note that the composition happens only internally on the server side. What I want, as it seems you understood, is to define this resource graph with "strongly typed" relationships and automatically have JsonApiFramework "decompose"/flatten/normalize that graph into the correct relationships and includes.

**Comment by scott-mcdonald** [comment]
Yes sir I completely understand what you are asking for. You want to define your service model or schema with an object graph where resources have relationships to other resources via aggregation/composition similar to Entity Framework for example. Then upon json serialization/deserialization the aggregated/composed related resources are "flattened" per the json:api specification and linked together what the json:api specification calls "resource linkage".

For JsonApiFramework, I designed it to be tightly coupled with the specification in terms of semantics and how you build the json:api document so relationships via composition was not considered, etc. What you are asking for would be an awesome feature but...

I took a step back from json:api because I don't like the slowness in which the specification has been evolving. For example I submitted a feature enhancement about [adding resource "version" as a top level](https://github.com/json-api/json-api/issues/600) that was debated and accepted for json:api 1.1 but has been sitting as a pull request for since 2015. Another example is "side posting" where the current json:api architectural constraint of only creating a single resource at a time is not pragmatic so there has been a proposal about being able to create primary and related resources in a single json:api document that seems to just sit there and not go anywhere...

Anyways I started looking at GraphQL in comparison to json:api and liked some of the concepts there but also love json:api as a hypermedia API specification. Then I came up with an idea...

I don't want to be tightly coupled to just "one" specification, so I decided what I really would like to develop is a specification agnostic framework that steals a lot of the SOLID principles and design patterns I used in JsonApiFramework for this endeavor. The core idea is to develop a specification agnostic service model (schema) with the "document" being a DOM (document object model) that based upon a "provider" implementation can serialize/deserialize the DOM for a given schema into something like a json:api representation or a GraphQL representation or whatever representation. I am calling this ApiFramework for now...

Now I will still be using JsonApiFramework at work for the foreseeable future but any extra time I have will be going into ApiFramework which will support the feature you are asking about directly. In other words I love your feature request and would like it myself but since it is non trivial I wouldn't have the time to add it. Non trivial features or bug fixes I would consider adding to JsonApiFramework though...
